### PR TITLE
Tierd file access for derived datasets

### PR DESCRIPTION
**Pull request by jgrosseo**

### CODE REVIEW

**Comment by jgrosseo** [comment]
@ktf this is a demonstrator how tierd file access could look like. This looks for not found trees in the parent file (the file from which the derived data has been produced) if such a file is provided in the map of the file. I have tested this with a locally patched input file and it works well for HF. 
Missing:
- monitoring the reading from the multiple files
- writing the map when derived data is produced
- adjusting the merger to merge the map

**Comment by jgrosseo** [comment]
This version is significantly refactored:
- DataInputDirector has been moved to the plugin
- The tree reading is in the DataInputDirector as well as the monitoring
- This allows recursive lookup of trees in parent files and monitoring info for each of the files

There are a few things which are still ugly and can be improved:
- Passing of quite some variables from the plugin into the DataInput* classes. Some by reference. Is the monitoring info about compressed/uncompressed size needed? 

What is missing:
- writing the map when derived data is produced
- adjusting the merger to merge the map
- adding a configurable allowing to replace fractions of the absolute parent file path to allow relocation of files to a local cluster

**Comment by jgrosseo** [comment]
The monitoring lines have now changed in the following way:

Old: 
`Read info: lfn=input/0/AnalysisResults_trees.root,size=39393,total_df=4,read_df=4,read_bytes=35721,read_calls=52,io_time=1.5,wait_time=0.1`

New: 
`Read info: lfn=input/0/AnalysisResults_trees.root,size=39393,total_df=4,read_df=4,read_bytes=35721,read_calls=52,io_time=0.0,wait_time=1.5,level=0
Read info: lfn=/home/jgrosseo/o2/hf/processSkim/input/0/AO2D.root,size=151126836,total_df=4,read_df=4,read_bytes=118399936,read_calls=543,io_time=1.4,wait_time=0.1,level=1
`
In case of parent file access, each file gets its own line. The `io_time` is properly split between the files. The `wait_time` can lead to double counting because each file can wait for others. Therefore the additional tag `level`, we should then only consider for the monitoring on AliEn the `wait_time` of the top-level entry...

File: Framework/AnalysisSupport/test/test_DataInputDirector.cxx [hunk 1]
```diff
@@ -16,7 +16,8 @@
 #include <boost/test/unit_test.hpp>
 
 #include "Headers/DataHeader.h"
-#include "Framework/DataInputDirector.h"
+// TODO how to write this better?
+#include "../src/DataInputDirector.h"
```
Comment (jgrosseo) [inline Framework/AnalysisSupport/test/test_DataInputDirector.cxx:19]: Somehow I don't manage to improve this even with `PRIVATE_INCLUDE_DIRECTORIES ${CMAKE_CURRENT_LIST_DIR}/src`
Could you advise?

Comment (ktf) [reply Framework/AnalysisSupport/test/test_DataInputDirector.cxx:19]: I think this is fine, frankly.

**Comment by jgrosseo** [comment]
This includes the writing of the map to the output file.

@ktf I need you to carefully check if the added routes do not mess up something for online/other use cases...

**Comment by jgrosseo** [comment]
Enabling the CI. Test on Hyperloop pending. Please do NOT merge, yet.

**Comment by jgrosseo** [comment]
The test failing in cs8 seems to be OK locally:
`        Start 101: test_Framework_test_SimpleDataProcessingDevice01
 87/106 Test #101: test_Framework_test_SimpleDataProcessingDevice01 ........   Passed    1.40 sec
`

**Comment by jgrosseo** [comment]
Tested on Hyperloop: OK.

**Comment by jgrosseo** [comment]
Everything is in now except:
- adjusting the merger to merge the map
which is anyway in O2Physics

**Comment by jgrosseo** [comment]
Last missing element: https://github.com/AliceO2Group/O2Physics/pull/1227

**Pull request merged**